src/librustc_mir/transform/promote_consts.rs

```diff
@@ -134,22 +157,644 @@ impl<'tcx> Visitor<'tcx> for TempCollector<'tcx> {
+            } => self.validate_local(*local),
+            PlaceRef {
+                base: PlaceBase::Static(box Static {
+                    kind: StaticKind::Promoted { .. },
+                    ..
+                }),
+                projection: [],
+            } => bug!("qualifying already promoted MIR"),
+            PlaceRef {
+                base: PlaceBase::Static(box Static {
+                    kind: StaticKind::Static,
+                    def_id,
+                    ..
+                }),
+                projection: [],
+            } => {
+                // Only allow statics (not consts) to refer to other statics.
+                // FIXME(eddyb) does this matter at all for promotion?
+                let allowed = self.is_static || self.is_static_mut;
+                if !allowed {
+                    return Err(Unpromotable);
+                }
+
+                let is_thread_local = self.tcx.has_attr(*def_id, sym::thread_local);
+                if is_thread_local {
+                    return Err(Unpromotable);
+                }
+
+                Ok(())
+            }
+            PlaceRef {
+                base: _,
+                projection: [proj_base @ .., elem],
+            } => {
+                match *elem {
+                    ProjectionElem::Deref |
+                    ProjectionElem::Downcast(..) => return Err(Unpromotable),
+
+                    ProjectionElem::ConstantIndex {..} |
+                    ProjectionElem::Subslice {..} => {}
+
+                    ProjectionElem::Index(local) => {
+                        self.validate_local(local)?;
+                    }
+
+                    ProjectionElem::Field(..) => {
+                        if self.is_non_const_fn {
+                            let base_ty =
+                                Place::ty_from(place.base, proj_base, self.body, self.tcx).ty;
+                            if let Some(def) = base_ty.ty_adt_def() {
+                                // No promotion of union field accesses.
+                                if def.is_union() {
+                                    return Err(Unpromotable);
+                                }
+                            }
+                        }
+                    }
+                }
+
+                self.validate_place(PlaceRef {
+                    base: place.base,
+                    projection: proj_base,
+                })
+            }
+        }
+    }
+
+    fn validate_operand(&self, operand: &Operand<'tcx>) -> Result<(), Unpromotable> {
+        match operand {
+            Operand::Copy(place) |
+            Operand::Move(place) => self.validate_place(place.as_ref()),
+
+            Operand::Constant(constant) => {
+                if let ConstValue::Unevaluated(def_id, _) = constant.literal.val {
+                    if self.tcx.trait_of_item(def_id).is_some() {
+                        // Don't peek inside trait associated constants.
```

Comment:
shouldn't these be unpromotable then? or at least make the promotability depend on the type?

Reply:
Promotability isn't type-based (the `NeedsDrop`/`HasInteriorMut` qualifs are checked elsewhere), the comment is misleading, I'll add some explanation.